pinot-core/src/main/java/com/linkedin/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -33,318 +38,292 @@
+          // Aggregation only query.
+          attachAggregationResults(brokerResponseNative, aggregationsInfo, dataTableMap);
         } else {
-          AggregationGroupByOperatorService aggregationGroupByOperatorService =
-              new AggregationGroupByOperatorService(brokerRequest.getAggregationsInfo(), brokerRequest.getGroupBy());
-          brokerResponseNative.setAggregationResults(
-              reduceOnAggregationGroupByOperatorResults(aggregationGroupByOperatorService, dataTableMap));
+          // Aggregation group-by query.
+          attachGroupByResults(brokerResponseNative, aggregationsInfo, brokerRequest.getGroupBy(), dataTableMap);
         }
-        return brokerResponseNative;
       }
-    } catch (Exception e) {
-      QueryProcessingException processingException = new QueryProcessingException();
-      processingException.setMessage(e.getMessage());
-      processingException.setErrorCode(QueryException.BROKER_GATHER_ERROR_CODE);
-      brokerResponseNative.getProcessingExceptions().add(processingException);
-      return brokerResponseNative;
     }
 
-    throw new UnsupportedOperationException(
-        "Should not reach here, the query has no attributes of selection or aggregation!");
+    return brokerResponseNative;
   }
 
   /**
-   * Reduce selection results from various servers into SelectionResults object, that goes
-   * into BrokerResponseNative.
+   * Given a data schema, remove data tables that do not match this data schema.
    *
-   * @param brokerRequest
-   * @param instanceResponseMap
-   * @return
+   * @param dataSchema data schema.
+   * @param dataTableMap map from server to data table.
+   * @return list of server names where the data table got removed.
    */
-  private SelectionResults reduceOnSelectionResults(BrokerRequest brokerRequest,
-      Map<ServerInstance, DataTable> instanceResponseMap) {
-    try {
-      if (instanceResponseMap.size() > 0) {
-        DataTable dt = chooseFirstNonEmptySchema(instanceResponseMap.values());
-        removeConflictingResponses(dt.getDataSchema(), instanceResponseMap);
-
-        if (brokerRequest.getSelections().isSetSelectionSortSequence()) {
-          SelectionOperatorService selectionService =
-              new SelectionOperatorService(brokerRequest.getSelections(), dt.getDataSchema());
-          return selectionService.renderSelectionResultsWithOrdering(selectionService.reduceWithOrdering(instanceResponseMap));
-        } else {
-          Collection<Serializable[]> reduceResult =
-              SelectionOperatorUtils.reduceWithoutOrdering(instanceResponseMap, brokerRequest.getSelections().getSize());
-
-          return SelectionOperatorUtils
-              .renderSelectionResultsWithoutOrdering(reduceResult, brokerRequest.getSelections().getSelectionColumns(),
-                  dt.getDataSchema());
-        }
-      } else {
-        return null;
+  private List<String> removeConflictingResponses(@Nonnull DataSchema dataSchema,
```

Comment:
What happens when the conflict is resolvable, eg: a column is long in one dataTable and int in another? We should be able handle such cases by treating both as long.